dsgrid/dataset/dataset_schema_handler_base.py

```diff
@@ -268,17 +236,33 @@ def _map_and_reduce_dimension(self, df, column, records):
         return df
 
     @staticmethod
+    @track_timing(timer_stats_collector)
     def _check_null_value_in_unique_dimension_rows(dim_table):
-        dim_with_null = {}
-        cols_to_check = {x for x in dim_table.columns if x != "id"}
-
-        for col in cols_to_check:
-            null_count = dim_table.filter(F.col(col).isNull()).count()
-            if null_count > 0:
-                dim_with_null[col] = null_count
+        if os.environ.get("__DSGRID_SKIP_NULL_UNIQUE_DIMENSION_CHECK__"):
+            # This has intermittently caused GC-related timeouts for TEMPO.
+            # Leave a backdoor to skip these checks, which may eventually be removed.
+            logger.warning("Skip _check_null_value_in_unique_dimension_rows")
+            return
 
-        if len(dim_with_null) > 0:
-            raise DSGInvalidDimensionMapping(
-                "Invalid dimension mapping application. "
-                f"Combination of remapped dataset dimensions contain NULL value(s) for dimension(s): \n{dim_with_null}"
-            )
+        cols_to_check = {x for x in dim_table.columns if x != "id"}
+        cols_str = ", ".join(cols_to_check)
+        filter_str = " OR ".join((f"{x} is NULL" for x in cols_to_check))
+        dim_table.createOrReplaceTempView("dim_table")
+
+        try:
+            # Avoid iterating with many checks unless we know there is at least one failure.
+            nulls = sql(f"SELECT {cols_str} FROM dim_table WHERE {filter_str}")
+            if not nulls.rdd.isEmpty():
+                dims_with_null = set()
+                for col in cols_to_check:
+                    if not nulls.select(col).filter(f"{col} is NULL").rdd.isEmpty():
+                        dims_with_null.add(col)
+                assert dims_with_null, "Did not find any dimensions with NULL values"
+
+                raise DSGInvalidDimensionMapping(
+                    "Invalid dimension mapping application. "
+                    "Combination of remapped dataset dimensions contain NULL value(s) for "
+                    f"dimension(s): \n{dims_with_null}"
+                )
+        finally:
+            sql("DROP VIEW dim_table")
```

Comment:
@lixiliu This reduced the time to process comstock in half. And it seems to prevent the GC error when running locally.

Reply:
It's cut in half because the code didn't get past `if not nulls.rdd.isEmpty()` for ComStock right? Though the for-loop should be faster now when prompted since it would be looping through only the null-containing subset.

Does it make sense to drop view right after `nulls` is created and before the for-loop since it's not used again?

Reply:
I ran a full test with all three datasets. The time actually dropped from 987s to 373s. The main issue is that starting a single Spark job has high overhead. The for loop triggers that overhead many times. This illustrates the importance of squashing things into one job even when it's harder to write.

The `nulls` dataframe is built on top of `dim_table`, so I'm not sure what would happen I dropped it before the for loop. To me it makes sense visually to see that `dim_table` is used across the code block. I doubt there would be any performance or memory difference.

One possible objection is that this for loop isn't tested. I don't know how to create a scenario where this would be entered. Any ideas?

Reply:
Don't drop null in _map_and_reduce() might work for ComStock since project only calls for a subset of ComStock geography.

Reply:
Yeah, I meant testing it without hacking the code. Testing once is different than testing at every invocation of pytest. The only way we have to do it with current infrastructure is to manipulate the test data. I've seen test infrastructure that allows arbitrarily changing values at runtime, but not with Python. I did perform the one-time test.

Reply:
Agreed with the test limitations, I don't think we can hack the test data. The null check is meant to catch errors when mapping/data reduction are not applied correctly, so it's checking for internal consistency in the code, not the data.